### PR TITLE
Optimize cross offers

### PR DESCRIPTION
**Pull request by jonjove**
Preliminary attempt to optimize `convertWithOffers` by changing the interface of `LedgerTxn::loadBestOffer`.

### CODE REVIEW

File: src/transactions/OfferExchange.cpp [hunk 1]
```diff
@@ -1106,10 +1106,24 @@ convertWithOffers(
...
+        if (offerTrail.empty())
+        {
+            wheatOffer = ltx.loadBestOffer(sheep, wheat);
+        }
+        else
+        {
+            LedgerKey key(OFFER);
+            key.offer().sellerID = offerTrail.back().sellerID;
+            key.offer().offerID = offerTrail.back().offerID;
+            wheatOffer = ltx.loadBestOffer(sheep, wheat, key);
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/transactions/OfferExchange.cpp]: Actually, I think the way the code was written before is a lot easier to understand.

I don't think we ever enumerate the order book without actually consuming offers.
If there is a way to do this (in older versions of the protocol), we should instead just deal with it in a different way (as in: if loadBestOffer returns the same offer than the last iteration, we have to stop iterating somehow); that way we can drop the support for this "exclude" parameter that is confusing.

File: src/ledger/LedgerTxn.cpp [hunk 2]
```diff
@@ -991,18 +1080,26 @@ LedgerTxn::Impl::loadAllOffers(LedgerTxn& self)
...
-                               Asset const& selling)
+                               Asset const& selling,
+                               LedgerKey const* previousBest)
 {
     throwIfSealed();
     throwIfChild();
 
     std::unordered_set<LedgerKey> exclude;
-    auto le = getBestOffer(buying, selling, exclude);
+    auto le = getBestOffer(buying, selling, previousBest);
```
Comment (MonsieurNicolas) [inline src/ledger/LedgerTxn.cpp]: `exclude` is not used anymore

File: src/ledger/LedgerTxn.cpp [hunk 3]
```diff
@@ -991,18 +1080,26 @@ LedgerTxn::Impl::loadAllOffers(LedgerTxn& self)
 LedgerTxnEntry
 LedgerTxn::loadBestOffer(Asset const& buying, Asset const& selling)
 {
-    return getImpl()->loadBestOffer(*this, buying, selling);
+    return getImpl()->loadBestOffer(*this, buying, selling, nullptr);
+}
+
+LedgerTxnEntry
+LedgerTxn::loadBestOffer(Asset const& buying, Asset const& selling,
+                         LedgerKey const& previousBest)
```
Comment (MonsieurNicolas) [inline src/ledger/LedgerTxn.cpp]: `previousBest` should just be a `LedgerKey const* previousBest` as you can deal with null

File: src/database/Database.cpp [hunk 4]
```diff
@@ -222,6 +222,11 @@ Database::applySchemaUpgrade(unsigned long vers)
         // add tracking table information
         mApp.getHerderPersistence().createQuorumTrackingTable(mSession);
         break;
+    case 11:
+        mSession << "DROP INDEX IF EXISTS bestofferindex;";
```
Comment (MonsieurNicolas) [inline src/database/Database.cpp]: why do we need this index? I thought previous benchmarks demonstrated that they were not needed

File: src/ledger/LedgerTxn.cpp [hunk 5]
```diff
@@ -1543,7 +1543,6 @@ LedgerTxnRoot::Impl::commitChild(EntryIterator iter, LedgerTxnConsistency cons)
     // Clearing the cache does not throw
     mBestOffersCache.clear();
     mEntryCache.clear();
-    mPrefetchMetrics.clear();
```
Comment (MonsieurNicolas) [inline src/ledger/LedgerTxn.cpp]: this seems to belong to a different PR

File: src/ledger/LedgerTxn.cpp [hunk 6]
```diff
@@ -1670,7 +1669,15 @@ LedgerTxnRoot::prefetch(std::unordered_set<LedgerKey> const& keys)
 uint32_t
 LedgerTxnRoot::Impl::prefetch(std::unordered_set<LedgerKey> const& keys)
 {
-    uint32_t total = 0;
+    size_t const maxSize = ENTRY_CACHE_FILL_RATIO * mEntryCache.maxSize();
```
Comment (MonsieurNicolas) [inline src/ledger/LedgerTxn.cpp]: this also belongs to a different PR (prefetch updates)

File: src/ledger/LedgerTxn.cpp [hunk 7]
```diff
@@ -496,68 +550,189 @@ LedgerTxn::Impl::getAllOffers()
...
+{
+    return getImpl()->getBestOffer(buying, selling, exclude);
+}
+
+std::shared_ptr<LedgerEntry const>
+LedgerTxn::Impl::getBestOffer(Asset const& buying, Asset const& selling,
+                              LedgerKey const* exclude)
+{
+    auto const& gbos = mGetBestOffersState;
+    if (!gbos && (mChild || !mEntry.empty()))
```
Comment (MonsieurNicolas) [inline src/ledger/LedgerTxn.cpp]: Isn't this just `if (!gbos)` ? I don't think it's valid to call `getBestOffer` without calling `prepareGetBestOffer` ?

Comment (jonjove) [reply src/ledger/LedgerTxn.cpp]: Actually, the most efficient way to use `prepareGetBestOffer` and `getBestOffer` is to:
1. Call `prepareGetBestOffer` on a `LedgerTxn`
2. Open a child `LedgerTxn`
3. Call `loadBestOffer` on the child (_without calling_ `prepareGetBestOffer` on the child)
4. Do things other than `loadBestOffer`
4. Commit the child
5. Repeat steps 2-4 as necessary

Indeed, this is what we do in `convertWithOffers`.

Why should we not call `prepareGetBestOffer` on the child? Because `prepareGetBestOffer` always recursively resets the streams of all parents to the beginning, so this will lead to `O(n^2)` iterations when making `O(n)` calls to `prepareGetBestOffer` followed by `getBestOffer`.

Why is it safe to not call `prepareGetBestOffer` when `!mChild && mEntry.empty()`? Because in this case there are no offers in the `LedgerTxn` so the `GetBestOfferState` won't contain anything. Note this argument only works if `prepareGetBestOffer` was already called for the parent. Technically I don't even think that the `!mChild` part is required.

Despite the fact that this works, it is neither beautiful nor intuitive. Any ideas for making this part of the interface cleaner would be much appreciated.

Comment (MonsieurNicolas) [reply src/ledger/LedgerTxn.cpp]: Ah yes, mmm, this is annoying. At a minimum it would be nice to enforce that `prepareGetBestOffer` was called on one of the parents. Right now I don't see how we can do this though

File: src/transactions/test/CrossOffersBenchmark.cpp [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+    std::vector<TransactionFramePtr> txs;
+    size_t nOps = 0;
+
+    uint64_t seqNum = getNextSeqNum(app, src);
+    for (auto iter = dest.cbegin(); iter != dest.cend();)
+    {
+        txs.emplace_back(
+            createAccountsTx(app, src, seqNum++, balance, iter, dest.cend()));
+        nOps += txs.back()->getEnvelope().tx.operations.size();
+        if (nOps + 100 > 5000)
```
Comment (vogel) [inline src/transactions/test/CrossOffersBenchmark.cpp]: two magic numbers here, could be replaced by constant and variable

**Comment by jonjove** [comment]
Replaced by #2307.
